### ISSUE TITLE
search filter `archived`/`fork` "yes/no/only" values are confusing

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#7003]
A user was confused by the fact that `archived:yes` turned up repositories when there were in fact no repositories on their Sourcegraph instance. In fact, it just "includes" archived repositories and `archived:only` shows no repositories as expected.

We should rename these somehow to make this more clear.

**Comment by keegancsmith** [comment]
Any proposals? It would be nice if it was obvious, and could somehow extend to other repo metadata. I suppose the difficult part is all our operators are joined by `AND`, except when you magically do `archived:yes` or `forked:yes` and instead it is grows the universe of repos.

Maybe instead of archived:yes, we have a single specifier which indicates _all_ repositories and then the user can remove repos from there. Or the default state somehow indicates it is a reduced set of repos (which may be useful for things like Sourcegraph.com). Either way this is quite tricky to do nicely :) Also cc @rvantonder

**Comment by rvantonder** [comment]
> A user was confused by the fact that archived:yes turned up repositories when there were in fact no repositories on their Sourcegraph instance

Just to clarify, do you mean no _archived_ repositories, or no repositories at all (how is the latter possible?)

> confusing

My impression is that the confusion stems from the fact that the type here is quasi-boolean value (`yes`/`no`), but actually performs inclusion/exclusion filter (`only`) on a set. I.e., it's hard to reconcile the type with the semantics/search behavior. Assuming I had a blank slate, I'd probably add a type that corresponds more closely to the search behavior. I think this gets close to the idea of having a single specifier that @keegancsmith talks about.

Without thinking at first directly about syntax, a good way to organize this is to say that we have a field or keyword "repos" (forget for the moment our current regex filter for the  repos field) and then a predicate language that applies to that field. In this predicate language, we define atomic values that are valid for `repos`, for example: `archived`, `forked`, `all`, `default` denoting sets of repos. The predicate language can then add expressions for negation, or set exclusion, etc., and define values in terms of others. For example, `default` might be `all \ (archived, forked)` (the set of all repos excluding archived and forked repos). I think working out a nice syntax would work well here, but will take time. 

To approximate a solution/syntax for the immediate issue, something like:

`filterRepos:archived`, `filterRepos:forked`, `filterRepos:all`, `filterRepos:default` could work. I'm not overly fond of the name `filterRepos` but it's the best I came up with in a couple of minutes. We can later expand this to the predicate language above. Note that I'm avoiding the idea of an `includeRepos` or `excludeRepos` filter since this just doesn't give enough precision for the semantics of things like `archived:only`. By `filterRepos:` I mean that it implies "only the values in this set".

**Comment by keegancsmith** [comment]
Yeah I like the idea of a more expressive language, especially if everything is still straightforward via our UI. An example of a language I used recently that I liked was gcloud's filter language https://cloud.google.com/sdk/gcloud/reference/topic/filters

But I almost find the hardest part in this is just naming the damn filters, while still keeping the good names for our most common uses.

**Comment by slimsag** [comment]
> Just to clarify, do you mean no archived repositories

Yes, apologies, I meant no archived repositories.

I think everything you both have said about the confusion being around the filter naming and this being a quasi-boolean value are correct. After filing this issue, the user also added this:

> all of the combinations are:
>
> ```
> fork:no archived:no
> fork:no archived:yes
> fork:no archived:only
> fork:yes archived:no
> fork:yes archived:yes
> fork:yes archived:only
> fork:only archived:no
> fork:only archived:yes
> fork:only archived:only
> ```
>
> I'd say `yes` does not behave as one would "intuitively expect", `only` behaves like I would have guessed `yes` does, perhaps if we rename `yes` to `any` (and `only` to `yes`) it would be more guessable

Now, I don't think we can do a renaming (that would break existing queries), @rvantonder's `filterRepos` suggestion is compelling.

It is also weird that:

1. You can say `fork:only archived:only` -- what does that actually mean? I guess "include ONLY forks AND include ONLY archived"?
2. `archived` is not `archive`, `fork` is not `forked` -- they don't match in naming either.

**Comment by rvantonder** [comment]
It's a bit of a misnomer. More accurate would be `selectRepos` or `pickRepos`.

**Comment by slimsag** [comment]
I have realized that what we do in #7004 will have some big effects here. For example, if we go with `selectRepos:` (or similarly named) proposed by Rijnard above:

```
selectRepos:archived
selectRepos:forked
selectRepos:all
selectRepos:default
```

And a site admin wants to change the `default` to include all forks and archives instead of excluding them by default (say because they have a heavy forking culture in their org), then you could not use `selectRepos` to exclude forks and archives selectively with just the four values above.

This makes me wonder if `include:forked|archived|all|default` and `exclude:forked|archived|all|default` might be good:

```
fork:no archived:no     => include:default
fork:no archived:yes    => include:archived
fork:no archived:only   => invalid
fork:yes archived:no    => include:forked
fork:yes archived:yes   => include:forked include:archived
fork:yes archived:only  => invalid
fork:only archived:no   => exclude:all include:forked
fork:only archived:yes  => invalid
fork:only archived:only => invalid
```

The default inclusion could be made configurable in global/user settings to solve #7004

This has a nice property that if you change the default to e.g. include forks and archives always, you just write `exclude` instead:

```
fork:no archived:no     => exclude:forked exclude:archived
fork:no archived:yes    => exclude:forked
fork:no archived:only   => invalid
fork:yes archived:no    => exclude:archived
fork:yes archived:yes   => exclude:default
fork:yes archived:only  => invalid
fork:only archived:no   => exclude:all include:forked
fork:only archived:yes  => invalid
fork:only archived:only => invalid
```

**Comment by rvantonder** [comment]
One shortcoming in my proposal is that the values themselves are not expressive enough for the general solution without the predicate language. I'm still be in favor of the predicate language long term, because of the points below. But we really only need conjunction for this case I think, which we can have in our current language by just repeating the `select` keyword to selectively enable either `archive` or `fork`:

(1) `select:default` => all repos excluding `archive` and `forked`
(2) `select:default select:forked` => as in (1), but including `forked` (set union)
(3) `select:default select:archived` => as in (1), but including `archived`
(4) `select:all` => an alias for `select:default select:archived select:forked`

For the `include`/`exclude` solution, the mapping makes sense, but there's also the complexity of invalidating combinations of `include`/`exclude` filters that doesn't happen for free. Just being explicit about that. But my question around this solution, related to what I raised earlier:

> I'm avoiding the idea of an includeRepos or excludeRepos filter since this just doesn't give enough precision for the semantics of things like archived:only

What does `archived:only` look like in the mapping above? `exclude:all include:archived`? If so, that's OK around how to express this, but the original issue here was a user who wanted the `archived:only` behavior, and it's not so pretty (i.e., non-obvious) that this translates to `exclude:all include:archived`. Just an observation. (And what I meant by 'precise' above was the ability to express this behavior with a value rather than an expression). 

This is a tradeoff that favors the scenario where:

> a site admin wants to ... exclude forks and archives selectively 

and the cost of easily expressing things like `archive:only`, and I'm not sure that we are offering a natural solution for the latter behavior.

**Comment by rvantonder** [comment]
And to be exhaustive, naturally:
 `select:archived` => only archived
 `select:forked` => only forked

**Comment by keegancsmith** [comment]
I like `select` a lot. Note: The way `select` combines will be the first time we `or` expressions. Maybe that is what we need to do, is introduce an or? But introducing or raises questions about how it interacts with non-select expressions. So probably safest to just document the behaviour of `select` explicitly.

Also need a way to `exclude` since for examples repos can be both archived and forked, but you may only want non-forked archived repos: `select:archive -select:fork`.

**Comment by rvantonder** [comment]
> or expressions

Yeah that corresponds to set union over labeled repos

> how it interacts with non-select expressions... So probably safest to just document the behaviour of select explicitly.

As an initial/immediate thing I would think of this behavior _only_ in the context of repos (select does not apply to file filters or such) and as a label attached to a repo (i.e., disjoint from repo names that can be filtered via regex). As a longer term thing we can combine these properties (repo names, repo labels, ...) under a single query language.

> Also need a way to exclude ...

Good point, didn't immediately think of negation as an explicit operator. Corresponds to set subtraction, so it makes sense. 

---

The above also prompts "OK what if I want only repos that are labeled both archived and forked, and not only archived or only forked" (i.e., `and` or set intersection of repo labels). Easy enough to describe as an operation, but the initial syntax/or-conjunction that I used doesn't account for this use case. We definitely want a canonical representation with verbose operators like `and` and `or`, e.g.,

- `select:"forked or archived"`

and then optionally a translation to/from that for user convenience. This is a not-immediately-available solution. The little thought I put into this is that there are multiple way do distinguish this behavior syntactically:

- `select:default|forked`
- `select:default select:forked`
- `select:default,forked`
- `select:(default,forked)`
...

And this will leave room for `and` in some other convenience syntax.

Right now, I'd recommend the least invasive change, even erring on the side of limiting what can be expressed, so that it's easy to retrofit into a general language/syntax soon-ish. As a minimal start, we only need `and`, `or`, and `not`. Concretely for this issue it may mean supporting the `or` behavior, picking an interim syntax, documenting it, and integrating it proper once we've had a roundtable discussion on a general query language and syntax.

**Comment by christinaforney** [comment]
This is a fantastic discussion and I really like where this is heading. If we're going to make a substantial change to our syntax like this, especially if we're introducing concepts of & | !, I would like to push to think about the entire syntax set. Supporting AND and OR queries has been a long requested item on the roadmap, and it feels like we're getting to a place where we can prioritize this. 

I would like to find a way to reduce confusion and churn with the syntax for users, but we do need to make some updates here. Can the change proposed above happen independently of a broader discussion or will we be setting ourselves up for difficulties later?

**Comment by keegancsmith** [comment]
> If we're going to make a substantial change to our syntax like this, especially if we're introducing concepts of & | !, I would like to push to think about the entire syntax set.

Yes, I think a redesigned syntax is a good thing to think about even if we don't necessarily implement it straight away. It can be our north star for making future decisions / iterate towards an idealized syntax.

> Supporting AND and OR queries has been a long requested item on the roadmap, and it feels like we're getting to a place where we can prioritize this.

Note that there is a lot of work in each backend to support AND and OR. But for operations like which repos to search is isolated to one part of our codebase so it is easier to implement these concepts.

> I would like to find a way to reduce confusion and churn with the syntax for users, but we do need to make some updates here. Can the change proposed above happen independently of a broader discussion or will we be setting ourselves up for difficulties later?

I think the `select:` syntax can be independent, baring our concerns around how they combine. Either way, I think optimizing around how people think about repo selection makes a lot of sense and shouldn't be held back by making it consistent with how we combine file selection / text search matches / etc. My personal preference is going ahead with introducing the `select:` syntax, especially if we have some great user documentation written for it first to iron out anything we might have missed so far.

**Comment by christinaforney** [comment]
> Yes, I think a redesigned syntax is a good thing to think about even if we don't necessarily implement it straight away. It can be our north star for making future decisions / iterate towards an idealized syntax.

Agreed!

> Note that there is a lot of work in each backend to support AND and OR. But for operations like which repos to search is isolated to one part of our codebase so it is easier to implement these concepts.

That's really helpful to know, thanks!

> I think the select: syntax can be independent, baring our concerns around how they combine. Either way, I think optimizing around how people think about repo selection makes a lot of sense and shouldn't be held back by making it consistent with how we combine file selection / text search matches / etc. My personal preference is going ahead with introducing the select: syntax, especially if we have some great user documentation written for it first to iron out anything we might have missed so far.

As long as we think about how this will combine with an improved syntax in the future, I support this. I've added this ticket to the Core services roadmap section so we can discuss priority.

**Comment by rvantonder** [comment]
> Note that there is a lot of work in each backend to support AND and OR. 

Just echoing that the `AND`/`OR`/`NOT` behavior is straightforward to support (_selectively_ for the search keywords for which these make sense). The thing to discuss and pin down is "user facing syntax" <-> search model

>My personal preference is going ahead with introducing the select: syntax, especially if we have some great user documentation written for it first to iron out anything we might have missed so far.

Agree with this.

> As long as we think about how this will combine with an improved syntax in the future, I support this.

The best way to achieve this is to not introduce any new syntax. By syntax I mean our existing `field:value` notation and space-delimited keywords--let's keep using that. The implication is that we cannot immediately support both `AND` and `OR` behavior with this syntax. Adding a new search keyword like `select` isn't "new syntax" as much as a value corresponding to behavior. My intuition would be to support `OR` expressions implicitly by specifying multiple `select`s, e.g., `select:foo select:bar` since that seems to be the natural move (and we document this), and we have code that supports parsing multiple fields. With the redesign, we can decide exactly what we want.

**Comment by christinaforney** [comment]
Thanks @rvantonder - this all sounds great to me. We'll prioritize on Monday :)

**Comment by keegancsmith** [comment]
@rvantonder do you think you will have any time to tackle this in 3.13?

**Comment by rvantonder** [comment]
We decided to shift priorities and backlog this for 3.13 in the interest of #8306.

**Comment by uwedeportivo** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.13 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by tsenart** [comment]
@rvantonder: Is it worth pulling this into 3.14 or should we leave it backlogged?

**Comment by rvantonder** [comment]
@tsenart I'd like for this to be on the radar for 3.14 but not sure what the overall workload looks like yet. Let's add it until we can pin down planning in 3.14 and put it in the backlog if it looks like things don't pan out.

**Comment by rvantonder** [comment]
I'm overcommitted for 3.14 and this issue has the lowest priority, so I'm taking it off. It's important and I really want to implement it though, so maybe I'll make partial progress.

**Comment by tsenart** [comment]
@rvantonder: Please triage.

**Comment by rvantonder** [comment]
This is important but not urgent. Keeping it in the backlog until we expose basic `and`/`or` query functionality.

**Comment by unknwon** [comment]
Removing "team/cloud" since this can be solely accomplished by the search team.

**Comment by keegancsmith** [comment]
@sourcegraph/search-platform this would be related to any work we have around improving our query language.

**Comment by uwedeportivo** [comment]
ok 🙂

**Comment by keegancsmith** [comment]
@uwedeportivo 👋
